src/ClassDiscovery.php

```diff
@@ -2,126 +2,104 @@
 
 namespace Http\Discovery;
 
-use Puli\Discovery\Api\Discovery;
+use Http\Discovery\Exception\DiscoveryFailedException;
+use Http\Discovery\Exception\StrategyUnavailableException;
 
 /**
  * Registry that based find results on class existence.
  *
  * @author David de Boer <david@ddeboer.nl>
  * @author Márk Sági-Kazár <mark.sagikazar@gmail.com>
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
  */
 abstract class ClassDiscovery
 {
     /**
-     * @var GeneratedPuliFactory
+     * A list of strategies to find classes.
+     *
+     * @var array
      */
-    private static $puliFactory;
+    public static $strategies = [
+        Strategy\Puli::class,
+        Strategy\HttpClients::class,
+        Strategy\GuzzleFactory::class,
+        Strategy\DiactorosFactory::class,
+    ];
 
     /**
-     * @var Discovery
+     * Discovery cache to make the second time we use discovery faster.
+     *
+     * @var array
      */
-    private static $puliDiscovery;
+    private static $cache = [];
 
     /**
-     * @return GeneratedPuliFactory
-     */
-    public static function getPuliFactory()
-    {
-        if (null === self::$puliFactory) {
-            if (!defined('PULI_FACTORY_CLASS')) {
-                throw new \RuntimeException('Puli Factory is not available');
-            }
-
-            $puliFactoryClass = PULI_FACTORY_CLASS;
-
-            if (!class_exists($puliFactoryClass)) {
-                throw new \RuntimeException('Puli Factory class does not exist');
-            }
-
-            self::$puliFactory = new $puliFactoryClass();
-        }
-
-        return self::$puliFactory;
-    }
-
-    /**
-     * Sets the Puli factory.
+     * Finds a class.
      *
-     * @param object $puliFactory
+     * @param string $type
+     *
+     * @return string
+     *
+     * @throws DiscoveryFailedException
      */
-    public static function setPuliFactory($puliFactory)
+    protected static function findOneByType($type)
```

Comment:
Made this function protected simply because I can't find any reason for it to be public.